### ISSUE TITLE
Type variable scoping with higher order functions

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#4948]
### Discussed in https://github.com/microsoft/pyright/discussions/4936

<div type='discussions-op-text'>

<sup>Originally posted by **sylee957** April 10, 2023</sup>
This is also a derived issue from #4852
But the issue happens with functions that take functions.

```python3
from dataclasses import dataclass
from typing import Callable, Generic, TypeVar

A = TypeVar("A")
B = TypeVar("B")
C = TypeVar("C")
X = TypeVar("X")


@dataclass(frozen=True)
class Pair(Generic[A, B]):
    left: A
    right: B


def identity(x: A) -> A:
    return x


def first(f: Callable[[A], B]) -> Callable[[Pair[A, X]], Pair[B, X]]:
    def wrapped(pair: Pair[A, X]) -> Pair[B, X]:
        return Pair(f(pair.left), pair.right)

    return wrapped
```

And applying the same function twice does not pass the type check
```python3
def test_1(pair: Pair[Pair[A, B], C]) -> Pair[Pair[A, B], C]:
    return first(first(identity))(pair)
```

While making duplicate function passes the type check.
```python3
def first_1(f: Callable[[A], B]) -> Callable[[Pair[A, X]], Pair[B, X]]:
    return lambda pair: Pair(f(pair.left), pair.right)


def first_2(f: Callable[[A], B]) -> Callable[[Pair[A, X]], Pair[B, X]]:
    return lambda pair: Pair(f(pair.left), pair.right)


def test_2(pair: Pair[Pair[A, B], C]) -> Pair[Pair[A, B], C]:
    return first_1(first_2(identity))(pair)
```</div>

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.312, which I just published. It will also be included in the next insiders release of pylance.
